release-content/0.14/release-notes/13417_implement_the_full_set_of_sort_methods_on_QueryIter.md

```diff
@@ -1 +1,114 @@
+With the default [`sort`], the lens has to be fully [`Ord`], like with [`slice::sort`].
+If this is not enough, we also have the counterparts to the remaining 6 sort methods from [`slice`]!
+
+The generic lens argument works the same way as with [`Query::transmute_lens`]. We do not use filters, they are inherited from the original query.
+The [`transmute_lens`] infrastructure has some nice additional features, which allows for this:
+
+```rust
+fn handle_enemies(enemies: Query<(&Health, &Attack, &Defense, &Rarity)>) {
+    for enemy in enemies.iter().sort_unstable::<Entity>() {
+        work_with(enemy)
+    }
+}
+```
+
+Because we can add [`Entity`] to any lens, we can sort by it without including it in the original query!
+
+These sort methods are fully generic over mutability, so can be used on both [`Query::iter`] and [`Query::iter_mut`]! The rest of the of the iterator methods on [`Query`] do not support sorting.
+The sorts return [`QuerySortedIter`], itself an iterator, enabling the use of further iterator adapters on it.
+Further, [`QuerySortedIter`] implements [`DoubleEndedIterator`] while the initial [`QueryIter`] does not. As a consequence, an empty sort can be used to get a more powerful iterator type:
+
+```rust
+fn handle_enemies(enemies: Query<(&Health, &Attack, &Defense, &Rarity)>) {
+    // A reversible query iterator!
+    enemies.iter_mut().sort::<()>().rev();
+}
+```
+
+Additionally, these query iterator sorts offer a workaround for a restriction on slice sorts:
+
+```rust
+// Some `Component` that holds data in an `Arc`.
+// `StatisticsData` does not implement `Copy`.
+#[derive(Component, Clone, Deref, PartialEq, Eq, PartialOrd, Ord)]
+pub struct Statistics(pub Arc<StatisticData>)
+
+// Does not compile.
+fn show_stats(users: Query<(&User, &Statistics)>) {
+   let mut users: Vec<_> = users.iter().collect();
+   users.sort_by_key(|(_, stats)| *stats)
+   show(users)
+}
+
+// Compiles.
+fn show_stats_2(users: Query<(&User, &Statistics)>) {
+   let users = users.iter().sort::<&Statistics>();
+   show(users)
+}
+```
+
+In current Rust, we can not return references from the key extraction closure in [`slice::sort_by_key`]/[`slice::sort_by_cached_key`].
+This can become a headache when using non-[`Copy`] [`Component`]s.
+The new [`sort`]/[`sort_by`] methods can work around this by having the lensing be the "key extraction", removing the need for such a closure.
+
+Keep in mind that the lensing does add some overhead, so these query iterator sorts do not perform equally to a manual sort on average. However, this *strongly* depends on workload, so best test it yourself if relevant!
+The sorts themselves are not yet cached between system runs, and these methods only sort iterators, not underlying query storage.
+
+Note that query iteration might happen in a deterministic order for now, but that may change anytime over future releases.
```

Comment:
I think this should be cut.

Reply:
I cut the last 2 sentences. Some people care a lot about sort performance, so I think the mention of it should remain.
It is not some simple "sort sugar" like itertools, but that might not be clear.